state/machine_linklayerdevices.go

```diff
@@ -862,19 +862,39 @@ func (m *Machine) AllAddresses() ([]*Address, error) {
 	return allAddresses, nil
 }
 
-// AllDeviceSpaceAddresses returns the SpaceAddress representation of all known
-// addresses assigned to link-layer devices on the machine.
+// AllDeviceSpaceAddresses returns the SpaceAddress representation of
+// all known addresses assigned to link-layer devices on the machine.
+// These addresses are populated with sufficient detail to be accurately
+// sortable by network.SortAddresses.
 func (m *Machine) AllDeviceSpaceAddresses() (corenetwork.SpaceAddresses, error) {
+	subnets, err := m.st.AllSubnets()
+	if err != nil {
+		return nil, errors.Annotate(err, "retrieving subnets")
+	}
+
 	var allAddresses corenetwork.SpaceAddresses
 	callbackFunc := func(doc *ipAddressDoc) {
-		allAddresses = append(allAddresses, corenetwork.SpaceAddress{
+		addr := corenetwork.SpaceAddress{
 			MachineAddress: corenetwork.NewMachineAddress(
 				doc.Value,
 				corenetwork.WithConfigType(doc.ConfigMethod),
 				corenetwork.WithCIDR(doc.SubnetCIDR),
 				corenetwork.WithSecondary(doc.IsSecondary),
 			),
-		})
+		}
+
+		// If this is not a loopback device, attempt to
+		// set the space ID based on the subnet.
+		if doc.ConfigMethod != corenetwork.ConfigLoopback && doc.SubnetCIDR != "" {
+			for _, sub := range subnets {
```

Comment:
This could be optimized at a later stage, so that subnets are cached outside of the callback via `map[string]string` (`map[cidr]spaceID`) and then used as an if check, rather than a for loop each time.